In [1]:
import sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
from mlxtend.frequent_patterns import apriori, association_rules

**Association analysis shows which item is frequently purchased with other items.**

In [2]:
# !conda install --yes --prefix {sys.prefix} numpy

In [3]:
# conda install -c conda-forge mlxtend

In [4]:
data = pd.read_excel('SalesReport.xls')

WARNING *** OLE2 inconsistency: SSCS size is 0 but SSAT size is non-zero


In [5]:
data.describe()

Row ID   Postal Code         Sales     Quantity     Discount  \
count  9994.000000   9983.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55245.233297    229.858001     3.789574     0.156203   
std    2885.163629  32038.715955    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  57103.000000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            Profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000

In [6]:
data.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2017-152156 2017-11-08 2017-11-11    Second Class    CG-12520   
1       2  CA-2017-152156 2017-11-08 2017-11-11    Second Class    CG-12520   
2       3  CA-2017-138688 2017-06-12 2017-06-16    Second Class    DV-13045   
3       4  US-2016-108966 2016-10-11 2016-10-18  Standard Class    SO-20335   
4       5  US-2016-108966 2016-10-11 2016-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0     42420.0   South  FUR-BO-10001798        Furniture    Bookcases   
1     42420.0   South  FUR-CH-10000454        Furniture       Chairs   
2     90036.0    West  OFF-LA-10000240  Office Supplies       Labels   
3     33311.0   South  FUR-TA-10000577        Furniture       Tables   
4     33311.0   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9983 non-null   float64       
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [8]:
data['State'].unique()

array(['Kentucky', 'California', 'Florida', 'North Carolina',
       'Washington', 'Texas', 'Wisconsin', 'Utah', 'Nebraska',
       'Pennsylvania', 'Illinois', 'Minnesota', 'Michigan', 'Delaware',
       'Indiana', 'New York', 'Arizona', 'Virginia', 'Tennessee',
       'Alabama', 'South Carolina', 'Oregon', 'Colorado', 'Iowa', 'Ohio',
       'Missouri', 'Oklahoma', 'New Mexico', 'Louisiana', 'Connecticut',
       'New Jersey', 'Massachusetts', 'Georgia', 'Nevada', 'Rhode Island',
       'Mississippi', 'Arkansas', 'Montana', 'New Hampshire', 'Maryland',
       'District of Columbia', 'Kansas', 'Vermont', 'Maine',
       'South Dakota', 'Idaho', 'North Dakota', 'Wyoming',
       'West Virginia'], dtype=object)

In [9]:
basket_california = (data[data['State'] == "California"]
                .groupby(['Order ID', 'Product Name'])["Quantity"]
                .sum().unstack().reset_index().fillna(0)
                .set_index('Order ID'))

In [10]:
basket_new_york = (data[data['State'] == "New York"]
                .groupby(['Order ID', 'Product Name'])["Quantity"]
                .sum().unstack().reset_index().fillna(0)
                .set_index('Order ID'))

In [11]:
basket_texas = (data[data['State'] == "Texas"]
                .groupby(['Order ID', 'Product Name'])["Quantity"]
                .sum().unstack().reset_index().fillna(0)
                .set_index('Order ID'))

In [12]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

In [19]:
basket_sets = basket_california.applymap(encode_units)

frequent_itemsets = apriori(basket_sets, min_support=0.001, use_colnames=True)
rules_california = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

In [20]:
rules_california

antecedents  \
0        (Microsoft Natural Ergonomic Keyboard 4000)   
1  (Boston 16765 Mini Stand Up Battery Pencil Sha...   

                                         consequents  antecedent support  \
0  (Boston 16765 Mini Stand Up Battery Pencil Sha...            0.002938   
1        (Microsoft Natural Ergonomic Keyboard 4000)            0.001959   

   consequent support   support  confidence        lift  leverage  conviction  
0            0.001959  0.001959    0.666667  340.333333  0.001953    2.994123  
1            0.002938  0.001959    1.000000  340.333333  0.001953         inf

In [32]:
rules_california[rules_california['antecedents'] == {'Microsoft Natural Ergonomic Keyboard 4000'}]

antecedents  \
0  (Microsoft Natural Ergonomic Keyboard 4000)   

                                         consequents  antecedent support  \
0  (Boston 16765 Mini Stand Up Battery Pencil Sha...            0.002938   

   consequent support   support  confidence        lift  leverage  conviction  
0            0.001959  0.001959    0.666667  340.333333  0.001953    2.994123

**For example, Microsoft Natural Ergonomic Keyboard 4000 is frequently purchased with Boston 16765 Mini Stand Up Battery Pencil Sharpener in California.**

In [23]:
basket_sets = basket_new_york.applymap(encode_units)

frequent_itemsets = apriori(basket_sets, min_support=0.001, use_colnames=True)
rules_new_york = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

In [24]:
rules_new_york

antecedents  \
0        (Adams Phone Message Book, 200 Message Capacit...   
1               (#10 Gummed Flap White Envelopes, 100/Box)   
2        (Dixon Prang Watercolor Pencils, 10-Color Set ...   
3               (#10 Gummed Flap White Envelopes, 100/Box)   
4                             (Sauder Barrister Bookcases)   
...                                                    ...   
5429413  (Pressboard Covers with Storage Hooks, 9 1/2" ...   
5429414  (SimpliFile Personal File, Black Granite, 15w ...   
5429415  (Hon 4070 Series Pagoda Armless Upholstered St...   
5429416                                   (Apple iPhone 5)   
5429417  (Plantronics Cordless Phone Headset with In-li...   

                                               consequents  \
0               (#10 Gummed Flap White Envelopes, 100/Box)   
1        (Adams Phone Message Book, 200 Message Capacit...   
2               (#10 Gummed Flap White Envelopes, 100/Box)   
3        (Dixon Prang Watercolor Pencils, 10-Color Set ...   
4               (#10 Gummed Flap White Envelopes, 100/Box)   
...                                                    ...   
5429413  (Logitech LS21 Speaker System - PC Multimedia ...   
5429414  (Logitech LS21 Speaker System - PC Multimedia ...   
5429415  (Logitech LS21 Speaker System - PC Multimedia ...   
5429416  (Logitech LS21 Speaker System - PC Multimedia ...   
5429417  (Logitech LS21 Speaker System - PC Multimedia ...   

         antecedent support  consequent support   support  confidence  \
0                  0.005338            0.003559  0.001779    0.333333   
1                  0.003559            0.005338  0.001779    0.500000   
2                  0.005338            0.003559  0.001779    0.333333   
3                  0.003559            0.005338  0.001779    0.500000   
4                  0.001779            0.003559  0.001779    1.000000   
...                     ...                 ...       ...         ...   
5429413            0.003559            0.001779  0.001779    0.500000   
5429414            0.001779            0.001779  0.001779    1.000000   
5429415            0.001779            0.001779  0.001779    1.000000   
5429416            0.003559            0.001779  0.001779    0.500000   
5429417            0.001779            0.001779  0.001779    1.000000   

               lift  leverage  conviction  
0         93.666667  0.001760    1.494662  
1         93.666667  0.001760    1.989324  
2         93.666667  0.001760    1.494662  
3         93.666667  0.001760    1.989324  
4        281.000000  0.001773         inf  
...             ...       ...         ...  
5429413  281.000000  0.001773    1.996441  
5429414  562.000000  0.001776         inf  
5429415  562.000000  0.001776         inf  
5429416  281.000000  0.001773    1.996441  
5429417  562.000000  0.001776         inf  

[5429418 rows x 9 columns]

In [27]:
rules_new_york[rules_new_york['antecedents'] == {'Plastic Binding Combs'}]

antecedents  \
291     (Plastic Binding Combs)   
493     (Plastic Binding Combs)   
1805    (Plastic Binding Combs)   
1908    (Plastic Binding Combs)   
1947    (Plastic Binding Combs)   
...                         ...   
362839  (Plastic Binding Combs)   
362900  (Plastic Binding Combs)   
362962  (Plastic Binding Combs)   
428372  (Plastic Binding Combs)   
908598  (Plastic Binding Combs)   

                                              consequents  antecedent support  \
291             (Acme Preferred Stainless Steel Scissors)            0.003559   
493              (Array Parchment Paper, Assorted Colors)            0.003559   
1805                    (Home/Office Personal File Carts)            0.003559   
1908      (Howard Miller 16" Diameter Gallery Wall Clock)            0.003559   
1947    (Imation 8gb Micro Traveldrive Usb 2.0 Flash D...            0.003559   
...                                                   ...                 ...   
362839  (Prang Dustless Chalk Sticks, Xerox 232, Imati...            0.003559   
362900  (Prang Dustless Chalk Sticks, Xerox 232, Home/...            0.003559   
362962  (Prang Dustless Chalk Sticks, Xerox 232, Imati...            0.003559   
428372  (Prang Dustless Chalk Sticks, Xerox 232, Imati...            0.003559   
908598  (Prang Dustless Chalk Sticks, Xerox 232, Imati...            0.003559   

        consequent support   support  confidence   lift  leverage  conviction  
291               0.003559  0.001779         0.5  140.5  0.001767    1.992883  
493               0.003559  0.001779         0.5  140.5  0.001767    1.992883  
1805              0.001779  0.001779         0.5  281.0  0.001773    1.996441  
1908              0.001779  0.001779         0.5  281.0  0.001773    1.996441  
1947              0.003559  0.001779         0.5  140.5  0.001767    1.992883  
...                    ...       ...         ...    ...       ...         ...  
362839            0.001779  0.001779         0.5  281.0  0.001773    1.996441  
362900            0.001779  0.001779         0.5  281.0  0.001773    1.996441  
362962            0.001779  0.001779         0.5  281.0  0.001773    1.996441  
428372            0.001779  0.001779         0.5  281.0  0.001773    1.996441  
908598            0.001779  0.001779         0.5  281.0  0.001773    1.996441  

[66 rows x 9 columns]

**Plastic Binding Combs is frequently purchased with Acme Preferred Stainless Steel Scissors and others more in New York.**

In [21]:
basket_sets = basket_texas.applymap(encode_units)

frequent_itemsets = apriori(basket_sets, min_support=0.001, use_colnames=True)
rules_texas = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

In [22]:
rules_texas

antecedents  \
0                                            (Avery 499)   
1              (#10- 4 1/8" x 9 1/2" Recycled Envelopes)   
2                    (Wasp CCD Handheld Bar Code Reader)   
3              (#10- 4 1/8" x 9 1/2" Recycled Envelopes)   
4      (Eldon Expressions Desk Accessory, Wood Photo ...   
...                                                  ...   
43589  (Boston School Pro Electric Pencil Sharpener, ...   
43590                                       (Newell 348)   
43591  (Lunatik TT5L-002 Taktik Strike Impact Protect...   
43592                        (GBC Plastic Binding Combs)   
43593  (GBC Standard Recycled Report Covers, Clear Pl...   

                                             consequents  antecedent support  \
0              (#10- 4 1/8" x 9 1/2" Recycled Envelopes)            0.004107   
1                                            (Avery 499)            0.002053   
2              (#10- 4 1/8" x 9 1/2" Recycled Envelopes)            0.002053   
3                    (Wasp CCD Handheld Bar Code Reader)            0.002053   
4         (#10- 4 1/8" x 9 1/2" Security-Tint Envelopes)            0.004107   
...                                                  ...                 ...   
43589  (Global Geo Office Task Chair, Gray, Xerox 199...            0.004107   
43590  (Global Geo Office Task Chair, Gray, Xerox 199...            0.006160   
43591  (Global Geo Office Task Chair, Gray, Xerox 199...            0.002053   
43592  (Global Geo Office Task Chair, Gray, Xerox 199...            0.004107   
43593  (Global Geo Office Task Chair, Gray, Xerox 199...            0.008214   

       consequent support   support  confidence        lift  leverage  \
0                0.002053  0.002053    0.500000  243.500000  0.002045   
1                0.004107  0.002053    1.000000  243.500000  0.002045   
2                0.002053  0.002053    1.000000  487.000000  0.002049   
3                0.002053  0.002053    1.000000  487.000000  0.002049   
4                0.002053  0.002053    0.500000  243.500000  0.002045   
...                   ...       ...         ...         ...       ...   
43589            0.002053  0.002053    0.500000  243.500000  0.002045   
43590            0.002053  0.002053    0.333333  162.333333  0.002041   
43591            0.002053  0.002053    1.000000  487.000000  0.002049   
43592            0.002053  0.002053    0.500000  243.500000  0.002045   
43593            0.002053  0.002053    0.250000  121.750000  0.002037   

       conviction  
0        1.995893  
1             inf  
2             inf  
3             inf  
4        1.995893  
...           ...  
43589    1.995893  
43590    1.496920  
43591         inf  
43592    1.995893  
43593    1.330595  

[43594 rows x 9 columns]

In [28]:
rules_texas[rules_texas['antecedents'] == {'Cisco SPA112 2 Port Phone Adapter'}]

antecedents  \
726    (Cisco SPA112 2 Port Phone Adapter)   
736    (Cisco SPA112 2 Port Phone Adapter)   
738    (Cisco SPA112 2 Port Phone Adapter)   
740    (Cisco SPA112 2 Port Phone Adapter)   
742    (Cisco SPA112 2 Port Phone Adapter)   
4195   (Cisco SPA112 2 Port Phone Adapter)   
4201   (Cisco SPA112 2 Port Phone Adapter)   
4207   (Cisco SPA112 2 Port Phone Adapter)   
4231   (Cisco SPA112 2 Port Phone Adapter)   
4237   (Cisco SPA112 2 Port Phone Adapter)   
4243   (Cisco SPA112 2 Port Phone Adapter)   
11042  (Cisco SPA112 2 Port Phone Adapter)   
11056  (Cisco SPA112 2 Port Phone Adapter)   
11070  (Cisco SPA112 2 Port Phone Adapter)   
11098  (Cisco SPA112 2 Port Phone Adapter)   
20953  (Cisco SPA112 2 Port Phone Adapter)   

                                             consequents  antecedent support  \
726                            (Cisco SPA 502G IP Phone)            0.004107   
736    (DMI Arturo Collection Mission-style Design Wo...            0.004107   
738    (GE General Purpose, Extra Long Life, Showcase...            0.004107   
740    (Permanent Self-Adhesive File Folder Labels fo...            0.004107   
742    (Wireless Extenders zBoost YX545 SOHO Signal B...            0.004107   
4195   (Cisco SPA 502G IP Phone, GE General Purpose, ...            0.004107   
4201   (Cisco SPA 502G IP Phone, Permanent Self-Adhes...            0.004107   
4207   (Cisco SPA 502G IP Phone, Wireless Extenders z...            0.004107   
4231   (GE General Purpose, Extra Long Life, Showcase...            0.004107   
4237   (Wireless Extenders zBoost YX545 SOHO Signal B...            0.004107   
4243   (Wireless Extenders zBoost YX545 SOHO Signal B...            0.004107   
11042  (Cisco SPA 502G IP Phone, GE General Purpose, ...            0.004107   
11056  (Cisco SPA 502G IP Phone, GE General Purpose, ...            0.004107   
11070  (Cisco SPA 502G IP Phone, Wireless Extenders z...            0.004107   
11098  (Wireless Extenders zBoost YX545 SOHO Signal B...            0.004107   
20953  (Cisco SPA 502G IP Phone, GE General Purpose, ...            0.004107   

       consequent support   support  confidence    lift  leverage  conviction  
726              0.004107  0.002053         0.5  121.75  0.002037    1.991786  
736              0.004107  0.002053         0.5  121.75  0.002037    1.991786  
738              0.004107  0.002053         0.5  121.75  0.002037    1.991786  
740              0.002053  0.002053         0.5  243.50  0.002045    1.995893  
742              0.004107  0.002053         0.5  121.75  0.002037    1.991786  
4195             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
4201             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
4207             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
4231             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
4237             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
4243             0.002053  0.002053         0.5  243.50  0.002045    1.995893  
11042            0.002053  0.002053         0.5  243.50  0.002045    1.995893  
11056            0.002053  0.002053         0.5  243.50  0.002045    1.995893  
11070            0.002053  0.002053         0.5  243.50  0.002045    1.995893  
11098            0.002053  0.002053         0.5  243.50  0.002045    1.995893  
20953            0.002053  0.002053         0.5  243.50  0.002045    1.995893

**Cisco SPA112 2 Port Phone Adapter is frequently purchased with Cisco SPA 502G IP Phone and others more in Texas.**